graph/src/components/store.rs

```diff
@@ -245,7 +276,111 @@ where
             })
         });
 
-        StoreEventStream { source }
+        StoreEventStream::new(Box::new(source))
+    }
+
+    /// Reduce the frequency with which events are generated while a
+    /// subgraph deployment is syncing. While the given `deployment` is not
+    /// synced yet, events from `source` are reported at most every
+    /// `interval`. At the same time, no event is held for longer than
+    /// `interval`. The `StoreEvents` that arrive during an interval appear
+    /// on the returned stream as a single `StoreEvent`; the events are
+    /// combined by using the maximum of all sources and the concatenation
+    /// of the changes of the `StoreEvents` received during the interval.
+    pub fn throttle_while_syncing(
+        self,
+        store: Arc<impl Store>,
+        deployment: SubgraphDeploymentId,
+        interval: Duration,
+    ) -> StoreEventStreamBox {
+        // We refresh the synced flag every SYNC_REFRESH_FREQ*interval to
+        // avoid hitting the database too often to see if the subgraph has
+        // been synced in the meantime. The only downside of this approach is
+        // that we might continue throttling subscription updates for a little
+        // bit longer than we really should
+        static SYNC_REFRESH_FREQ: u32 = 4;
+
+        let mut synced = false;
+        let synced_check_interval = interval.checked_mul(SYNC_REFRESH_FREQ).unwrap();
+        let mut synced_last_refreshed =
+            Instant::now() - synced_check_interval - synced_check_interval;
+
+        let mut pending_event: Option<StoreEvent> = None;
+        let mut source = self.source.fuse();
+        let mut had_err = false;
+        let mut delay = tokio_timer::Delay::new(Instant::now() + interval);
+
+        let source = Box::new(poll_fn(move || -> Poll<Option<StoreEvent>, ()> {
+            if had_err {
+                // We had an error the last time through, but returned the pending
+                // event first. Indicate the error now
+                had_err = false;
+                return Err(());
+            }
+
+            if !synced && synced_last_refreshed + synced_check_interval < Instant::now() {
+                // The special 'subgraphs' subgraph is never considered
+                // synced so that we always throttle it
+                synced = deployment != *SUBGRAPHS_ID
+                    && store
+                        .is_deployment_synced(deployment.clone())
+                        .unwrap_or(false);
+                synced_last_refreshed = Instant::now();
+            }
+
+            if synced {
+                return source.poll();
+            }
+
+            // Check if interval has passed since the last time we sent something.
+            // If it has, start a new delay timer
+            let ready = match delay.poll() {
+                Ok(Async::NotReady) => false,
```

Comment:
Why have the `ready` value instead of immediately returning `NotReady` in this case?

Reply:
Because I also want to make progress on getting `StoreEvents` so they don't pile up in the queue for the entire interval. Basically, this is so that `poll` makes as much progress as it possibly can before returning `Async::NotReady`.

Reply:
That part is the only one that I feel is not obvious. Perhaps a longer comment about what's going on there would be good. Otherwise: nice job!

Reply:
And maybe `ready` is not the ideal name for the variable. `throttling_interval_elapsed` maybe?

Reply:
I see there actually is a comment at the top of the function, so only consider the variable naming comment. 😄